dynamic_array.c

```diff
@@ -0,0 +1,340 @@
+#include "dynamic_array.h"
+
+// NOTE: When we destroy items, we call the destructor for them 
+// iff the destructor is not NULL.
+
+typedef void (*destr)(void *);
+
+// Initialiser - sets fields in instance and allocates buffer (via buffer_alloc)
+void buffer_init(struct buffer *inst, size_t item_size, size_t capacity, 
+                 size_t alloc_by, void (*destructor)(void *))
+{   
+    inst->capacity   = 0;
+    inst->length     = 0;
+    inst->item_size  = item_size;
+    inst->alloc_by   = alloc_by;
+    inst->data       = NULL;
+    inst->destructor = destructor;
+    inst->buffer     = NULL;
+
+    buffer_alloc(inst, capacity);
+}
+
+// Destructor - frees the buffer, calling the destructor for each item in the buffer
+void buffer_destroy(struct buffer *inst)
+{
+    for(int i = 0; i < inst->capacity; i++){
+        ((destr)(inst->destructor))(inst->buffer[i]);
+    }
+    inst->length = 0;
+    free(inst->buffer);
+    inst = NULL;
+}
+
+// If realloc SUCCEDS we modify the size
+static void try_realloc(struct buffer *inst, size_t new_capacity)
+{
+    inst->buffer = realloc_zero(inst->buffer, 
+                                buffer_size(inst) * sizeof(void*),
+                                new_capacity * sizeof(void*));
+    for (size_t i = inst->capacity; i < new_capacity; ++i) {
+        void *succeds =  realloc(inst->buffer[i], inst->item_size);
+        if (succeds){
+            inst->buffer[i] = succeds;
+        } else {
+            exit(0);
+        }
+    }  
+    
+    inst->capacity = new_capacity;
+    if (inst->length > new_capacity) {
+        inst->length = inst->capacity;
+    }
+}
+
+// Allocator
+//  * Reallocates the buffer to a new_capacity, preserving existing items 
+//     (unless new buffer is too small)
+//  * Reduces length if new capacity is smaller than previous length
+//  * For new capacity of zero, free() the buffer and set the pointer to NULL
+//  * When reducing length, call destructor for any items which are being removed
+void buffer_alloc(struct buffer *inst, size_t new_capacity)
+{  
+    if (new_capacity == 0){
+        buffer_destroy(inst);
+    } else{
+        // realloc will free the necessary memory
+        try_realloc(inst, new_capacity);
+    }
+}
+
+// Grow the buffer if necessary, to ensure the requested minimum capacity
+void buffer_reserve(struct buffer *inst, size_t min_capacity)
+{
+    if(inst->capacity < min_capacity) {
```

Comment:
We could invert this condition:

    if (inst->capacity >= min_capacity) {
        return;
    }

Then the rest of the function will be less indented and over-all, more readable.